mirascope/chat/models/async_openai_chat.py

```diff
@@ -0,0 +1,160 @@
+        self.model = model
+
+    async def create(
+        self,
+        prompt: Optional[Union[Prompt, str]] = None,
+        tools: Optional[list[Union[Callable, Type[OpenAITool]]]] = None,
+        **kwargs,
+    ) -> OpenAIChatCompletion:
+        """Asynchronously Makes a call to the model using `prompt`.
+
+        Args:
+            prompt: The prompt to use for the call. This can either be a `Prompt`
+                instance, a raw string, or `None`. If `prompt` is `None`, then the call
+                will attempt to use the `messages` keyword argument.
+            tools: A list of `OpenAITool` types or `Callable` functions that the
+                creation call can decide to use. If `tools` is provided, `tool_choice`
+                will be set to `auto`.
+            **kwargs: Additional keyword arguments to pass to the API call. You can
+                find available keyword arguments here:
+                https://platform.openai.com/docs/api-reference/chat/create
+
+        Returns:
+            A `OpenAIChatCompletion` instance.
+
+        Raises:
+            ValueError: if neither `prompt` nor `messages` are provided.
+        """
+        if tools:
+            openai_tools: list[type[OpenAITool]] = [
+                tool if isclass(tool) else convert_function_to_openai_tool(tool)
+                for tool in tools
+            ]
+            kwargs["tools"] = [tool.tool_schema() for tool in openai_tools]
+            if "tool_choice" not in kwargs:
+                kwargs["tool_choice"] = "auto"
+
+        if not prompt:
+            if "messages" not in kwargs:
+                raise ValueError("Either `prompt` or `messages` must be provided.")
+            messages = kwargs.pop("messages")
+        else:
+            messages = get_openai_messages_from_prompt(prompt)
+
+        return OpenAIChatCompletion(
+            completion=await self.client.chat.completions.create(
+                model=self.model,
+                messages=messages,
+                stream=False,
+                **kwargs,
+            ),
+            tool_types=openai_tools if tools else None,
+        )
+
+    async def stream(
+        self,
+        prompt: Optional[Union[Prompt, str]] = None,
+        **kwargs,
+    ) -> AsyncGenerator[OpenAIChatCompletionChunk, None]:
+        """Asynchronously streams the response for a call to the model using `prompt`.
+
+        Args:
+            prompt: The `Prompt` to use for the call.
+            **kwargs: Additional keyword arguments to pass to the API call. You can
+                find available keyword arguments here:
+                https://platform.openai.com/docs/api-reference/chat/create
+
+        Yields:
+            A `OpenAIChatCompletionChunk` for each chunk of the response.
+
+        Raises:
+            Re-raises any exceptions thrown by the openai chat completions
+            when iterating through the generator.
+        """
+        if not prompt:
```

Comment:
please add a todo here to refactor all of the setup into a `setup_openai_kwargs` function or something when we add tools to streaming (since at that point all four of create/acreate/stream/astream will have exactly duplicated code.